distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py

```diff
@@ -0,0 +1,297 @@
+      return strategy.reduce(tf.distribute.ReduceOp.SUM, losses, axis=None)
+
+
+    @tf.function
+    def worker_eval_fn(iterator):
+
+      def eval_fn(inputs):
+        """Evaluation function"""
+        batch_data, labels = inputs
+        predictions = model(batch_data, training=False)
+        eval_accuracy.update_state(labels, predictions)
+
+      inputs = next(iterator)
+      strategy.run(eval_fn, args=(inputs,))
+
+    checkpoint_manager = tf.train.CheckpointManager(
+        tf.train.Checkpoint(model=model, optimizer=optimizer),
+        FLAGS.checkpoint_dir,
+        max_to_keep=2)
+    if checkpoint_manager.latest_checkpoint:
+      checkpoint = checkpoint_manager.checkpoint
+      checkpoint.restore(
+          checkpoint_manager.latest_checkpoint
+          ).assert_existing_objects_matched()
+
+    train_dataset_fn = lambda _: cifar_preprocessing.input_fn(
+        is_training=True,
+        data_dir=FLAGS.data_dir,
+        batch_size=BATCH_SIZE,
+        parse_record_fn=cifar_preprocessing.parse_record,
+        dtype=tf.float32,
+        drop_remainder=True)
+    eval_dataset_fn = lambda _: cifar_preprocessing.input_fn(
+        is_training=False,
+        data_dir=FLAGS.data_dir,
+        batch_size=EVAL_BATCH_SIZE,
+        parse_record_fn=cifar_preprocessing.parse_record,
+        dtype=tf.float32)
+
+    # The following wrappers will allow efficient prefetching to GPUs
+    # when GPUs are supported by ParameterServerStrategy
+    @tf.function
+    def per_worker_train_dataset_fn():
+      return strategy.distribute_datasets_from_function(train_dataset_fn)
+
+    @tf.function
+    def per_worker_eval_dataset_fn():
+      return strategy.distribute_datasets_from_function(eval_dataset_fn)
+
+    per_worker_train_dataset = coordinator.create_per_worker_dataset(
+      per_worker_train_dataset_fn)
+    per_worker_eval_dataset = coordinator.create_per_worker_dataset(
+      per_worker_eval_dataset_fn)
+
+  global_steps = int(optimizer.iterations.numpy())
+  logging.info("Training starts with global_steps = %d", global_steps)
+  current_time = datetime.now().strftime("%Y%m%d-%H%M%S")
+  train_log_dir = FLAGS.train_log_dir + current_time
+  train_summary_writer = tf.summary.create_file_writer(train_log_dir)
+
+  for epoch in range(global_steps // STEPS_PER_EPOCH,
+                     TRAIN_EPOCHS):
+    per_worker_train_iterator = iter(per_worker_train_dataset)
+    per_worker_eval_iterator = iter(per_worker_eval_dataset)
+    for _ in range(STEPS_PER_EPOCH):
+      coordinator.schedule(worker_train_fn, args=(per_worker_train_iterator,))
+    coordinator.join()
+    logging.info("Finished joining at epoch %d. Training accuracy: %f.",
+                  epoch, train_accuracy.result())
+
+    # Since we are running inline evaluation below, a side-car evaluator job is not necessary.
+    for _ in range(EVAL_STEPS_PER_EPOCH):
+      coordinator.schedule(worker_eval_fn, args=(per_worker_eval_iterator,))
```

Comment:
We can probably build a similar API for DTensor async training. A major difficulty to sort out is what to do if worker_eval_fn( and or replica_fn) is multi-mesh -- for example if there is a summary Op that needs to run on a the CPU.